doc/sphinx/proof-engine/tactics.rst

```diff
@@ -650,180 +649,339 @@ Applying theorems
+   The matching subterms are then substituted into the type of :n:`@one_term`
+   when generating the updated goal or hypothesis.  Unmatched premises become
+   new subgoals with similar substitutions.  If no match is found, the
+   tactic fails.
+
+   The goal and hypothesis cases are described separately for clarity.
+
+   :n:`@one_term`
+     If :n:`@one_term` is an :n:`@ident`, it is the name of
+     a theorem, lemma or hypothesis whose type is given in the
+     theorem statement or shown in the context.  Otherwise it is a proof term whose
+     type can be displayed with :cmd:`Check` :n:`@one_term`.
+
+   Without :n:`@in_hyp_as` (the goal case)
+     If the goal matches all of the type of :n:`@one_term` (both premises and
+     the conclusion), the tactic proves the goal.
+     Otherwise, the tactic matches the goal against the conclusion of :n:`@one_term`.
+     If the match succeeds, the tactic replaces the current goal with a subgoal for
+     each non-dependent premise of the type of :n:`@one_term`.  See the example
+     :ref:`here <apply_backward>`.
+
+     If the conclusion of the type of :token:`one_term` does
+     not match the goal *and* the conclusion is an inductive type isomorphic to
+     a tuple type, then each component of the tuple is recursively matched to
+     the goal in the left-to-right order.
+
+     .. todo PR: does the next paragraph apply to the hypothesis case, too?
+        Also, can one of you help me create an example?
+
+     The tactic uses first-order unification with dependent
+     types unless the target (the goal conclusion or hypothesis to match)
+     is of the form
+     :n:`P (t__1 … t__n)` with ``P`` to be instantiated. In the latter case,
+     the behavior depends on the form of the target. If the target is of the form
+     :n:`(fun x__1 … x__n => Q) u__1 … u__n` and the :n:`t__i` and :n:`u__i` unify,
+     then :g:`P` is instantiated into :n:`(fun x__1 … x__n => Q)`. Otherwise,
+     :tacn:`apply`
+     tries to define :g:`P` by abstracting over :n:`t__1 … t__n` in the target.
+     You can use :tacn:`pattern` to transform the target so that it
+     gets the form :n:`(fun x__1 … x__n => Q) u__1 … u__n`.
+
+   :n:`@in_hyp_as` (the hypothesis case)
+     Proceeding from *right to left*, find the first premise of the type of
+     :n:`@one_term` that matches the specified hypothesis.  If a match
+     is found, the hypothesis is replaced with the conclusion of the type of
+     :n:`one_term` (substituting for the unified variables)
+     and the tactic creates a new subgoal for each unmatched premise.
+     See the example :ref:`here <apply_forward>`.
+
+     If specified, :n:`as @simple_intropattern` is applied to the conclusion
+     of the type of :n:`@one_term`.
+
+     .. todo PR the following old text was under apply ... in <hyp>,
+        but the "match the goal" suggests it's for the goal case of apply.
+        Which is it?  Also, can one of you help me create an example?
```

Comment:
Indeed, this text is wrong (was probably copy-pasted when introduced [14 years ago](70af80aad166bc54e4bbc80dfc9427cfee32aae6), improperly adapted and never fixed!). According to the example that the doc already contains, it should say something like "If the type of one_term is an inductive type isomorphic to a tuple type, then the tuple is (recursively) decomposed and the first component of the tuple of which a non-dependent premise matches the conclusion of the hypothesis is used."

Here is the example already contained in the text transformed into actual Coq code:

```coq
Parameters A B C : Prop.

Goal (A <-> B) -> A -> False.
intros H1 H2.
apply H1 in H2.
```

Reply:
OK, but I don't understand how "isomorphic to a tuple type" fits in here.

Reply:
Tuple is a generalization of couple to any size: https://en.wikipedia.org/wiki/Tuple

An inductive type isomorphic to a tuple type means any inductive type with a single constructor of the form:

```coq
Inductive type_name :=
| constructor_name : first_arg_type -> second_arg_type -> ... -> last_arg_type -> type_name.
```

So yes, that also matches the (non-primitive) definition of records.

These are the inductive types on which the `split` tactic will succeed, by transforming a goal which matches `type_name` into one goal per tuple component (arguments of `constructor_name`).

`A <-> B` is defined as `(A -> B) /\ (B -> A)` and `/\` is an example of "inductive type isomorphic to a tuple type".

FWIW, I don't mind replacing the mouthful "inductive type isomorphic to a tuple type" by something else, but in this case we should replace it consistently everywhere.

Reply:
I know what a tuple is in math and OCaml, but not in Coq.    We don't even define "tuple" explicitly in the doc.  (FTR, I hadn't seen "couple" to mean a 2-tuple before.)

> An inductive type isomorphic to a tuple type means any inductive type with a single constructor of the form:

Then probably it would be more readable just to give that form.  The isomorphic tuple type is `(first_arg_type second_arg_type ... last_arg_type)`?  I don't even know the proper notation for a tuple type.

> So yes, that also matches the (non-primitive) definition of records.

AFAICT, fields in records have to have names.  Where are the names?

Reply:
> FWIW, I don't mind replacing the mouthful "inductive type isomorphic to a tuple type" by something else, but in this case we should replace it consistently everywhere.

Searching for "isomorphic", I only see 3 uses of the phrase.  Not sure what would be a good replacement.  I also see "If an inductive type is isomorphic to the boolean type".  It seems too vague for the doc to just say there is a one-one mapping without specifying that mapping and exactly what is preserved when you apply the mapping.  (Without knowing what's preserved, all you know is that the cardinality of the two domains is the same.)

Reply:
> (FTR, I hadn't seen "couple" to mean a 2-tuple before.)

That may be a Frenchism then. Probably the correct English wording is ordered pair.

>Then probably it would be more readable just to give that form.

Sounds good to me. This "isomorphic" wording is just a way of carrying an intuitive meaning, but saying "inductive type of the form..." is clearer.

>AFAICT, fields in records have to have names. Where are the names?

Indeed, so better not to talk about records (names are optional in the current context).

> It seems too vague for the doc to just say there is a one-one mapping without specifying that mapping and exactly what is preserved when you apply the mapping.

My understanding is that isomorphic in this context means same definition up to renaming the types and fields.

Reply:
This is not yet making much (any?) sense.

> An inductive type isomorphic to a tuple type means any inductive type with a single constructor of the form:

These are correct, right?:
- A tuple _is_ a record?
- Record fields are ordered?
- Field names are not relevant to determining isomorphism, right?

What does "decompose" mean below?  Is just a depth-first search for the first match?? Can you give example(s) of decomposing?  The recursion is to handle the case where the type of a record field has a record type?

> If the type of :n:`@one_term` is an inductive type isomorphic to a tuple type,
then the tuple is (recursively) decomposed and _the first component of the
tuple for which a premise matches the conclusion of the
hypothesis is used._  :ref:`Example <example_apply_hyp_inductive>`

The italic part doesn't make any sense; "for which a premise" is a premise of the hypothesis??  And exactly how is the tuple "used"?

For your example (below), what is the tuple type and where is it defined?

```coq
Parameters A B C : Prop.

Goal (A <-> B) -> A -> False.
intros H1 H2.
apply H1 in H2.
```

Reply:
`A <-> B` is actually `(A -> B) /\ (B -> A)`, which is isomorphic to a pair type. Regarding "for which a premise", it is about the component of the tuple, so the documentation seems correct. Here, the premise `A` of the component `A -> B` of the tuple `A <-> B` is matched against the hypothesis `H2: A`. In other words, `apply H1 in H2` behaves like `apply (proj1 H1) in H2`, since the first component matched.

I had no idea the search was actually recursive. A quick test shows that it is indeed the case. And it is indeed depth-first. But the documentation is wrong about "the first component" being used. It seems to be the last one, i.e., the rightmost one.

Reply:
I wonder whether the recursive search is really useful or just a curio.  If it is, we should get an example for that.

Reply:
`A /\ B /\ C` (resp. `A * B * C`), since it is implicitly `A /\ (B /\ C)` (resp. `(A * B) * C`). The mechanism would be painful to use if it was not right-recursive (resp. left-recursive). And once it is both left- and right-recursive, it might just as well be fully recursive.

Reply:
> But the documentation is wrong about "the first component" being used. It seems to be the last one, i.e., the rightmost one.

I think not.  For the following, it uses the leftmost one.  What is your example?

```coq
Goal forall (iff: ((2 = 2) <-> (1=1)) <-> ((1 = 1) <-> 3 = 3))
   (H: 1 = 1), 1 = 1.
intros iff H.
(*
1 subgoal
iff : (2 = 2 <-> 1 = 1) <-> (1 = 1 <-> 3 = 3)
H : 1 = 1
______________________________________(1/1)
1 = 1
*)

apply iff.
(*
2 subgoals
iff : (2 = 2 <-> 1 = 1) <-> (1 = 1 <-> 3 = 3)
H : 1 = 1
______________________________________(1/2)
1 = 1 <-> 3 = 3
______________________________________(2/2)
2 = 2
*)
```

Reply:
This is a curious but harmless oddity.  Not sure if it's seen in practice.  The second and third goals are identical--we don't need both:

```coq
Goal forall A B C D: Prop, (A -> B) -> (A -> B -> C) -> True.
intros A B C D H0 H1.
apply H1 in H0.
(*
3 subgoals
A, B, C, D : Prop
H0 : C
H1 : A -> B -> C
______________________________________(1/3)
True
______________________________________(2/3)
A
______________________________________(3/3)
A
*)
```

While this works a bit differently.  I think `((A -> B) -> C) -> (A -> (B -> C))` is probably not true:

```coq
Goal forall A B C D: Prop, (A -> B) -> ((A -> B) -> C) -> True.
intros A B C D H0 H1.
apply H1 in H0.
(*
1 subgoal
A, B, C, D : Prop
H0 : A -> B
H1 : (A -> B) -> C
______________________________________(1/1)
True
*)
```

Reply:
> I think not. For the following, it uses the leftmost one.

No, it uses the rightmost one. You can double-check that the goal you get is the same as the one you would get with `apply (proj2 iff)`, and different from the one you would get with `apply (proj1 iff)`.

Consider the following much simpler example, in which you are left with `B`:
```coq
Goal forall A B Z : Prop, (A -> Z) /\ (B -> Z) -> Z.
Proof. intros. apply H.
```

Reply:
Got it, thanks.  Like many things in Coq, you often have to visualize several substeps correctly to make sense of things.